kalite/coachreports/views.py

```diff
@@ -271,6 +274,130 @@ def tabular_view(request, facility, report_type="exercise"):
+
+    exercises = sorted(temp, key=lambda e: (e["h_position"], e["v_position"]))
+    context["exercises"] = exercises
+    exercise_count = len(exercises)
+
+    # More code, but much faster
+    exercise_names = [ex["name"] for ex in context["exercises"]]
+    # Get students
+    context["students"] = []
+    exlogs = ExerciseLog.objects \
+        .filter(user__in=users, exercise_id__in=exercise_names) \
+        .order_by(*["user__%s" % field for field in student_ordering]) \
+        .values("user__id", "struggling", "complete", "exercise_id", "attempts", "streak_progress")
+    exlogs = list(exlogs)  # force the query to be evaluated
+
+    exercise_ids = [ex["id"] for ex in context["exercises"]]
+    user_count = len(users)
+
+    exercise_stats = {}
+    for ex_id in exercise_ids:
+        exercise_stats[ex_id] = { "struggling" :0 , "mastered": 0, "progress": 0, "mastery": 0}
+
+    for ex in exlogs:
+        if ex["complete"] == True:
+            exercise_stats[ex["exercise_id"]]["mastered"] = exercise_stats[ex["exercise_id"]]["mastered"] + 1
+        elif ex["struggling"] == True:
+            exercise_stats[ex["exercise_id"]]["struggling"] = exercise_stats[ex["exercise_id"]]["struggling"] + 1
+        elif ex["attempts"] > 0:
+            exercise_stats[ex["exercise_id"]]["progress"] = exercise_stats[ex["exercise_id"]]["progress"] + 1
+
+        exercise_stats[ex["exercise_id"]]["mastery"] = exercise_stats[ex["exercise_id"]]["mastered"] * 100 / user_count
+
+    context["exercise_stats"] = exercise_stats
+    context["exercise_count"] = exercise_count
+
+    exlog_idx = 0
+    for user in users:
+        log_table = {}
+        while exlog_idx < len(exlogs) and exlogs[exlog_idx]["user__id"] == user.id:
+            exlogs[exlog_idx]["streak_progress"] = exlogs[exlog_idx]["streak_progress"] / 12
+            log_table[exlogs[exlog_idx]["exercise_id"]] = exlogs[exlog_idx]
+            exlog_idx += 1
+
+        progress = 0
+        mastered = 0
+        struggling = 0
+
+        for ex in log_table:
+            if log_table[ex]["complete"] == True:
+                mastered = mastered + 1
+            elif log_table[ex]["struggling"] == True:
+                struggling = struggling + 1
+            elif log_table[ex]["attempts"] > 0:
+                progress = progress + 1
+
+        mastery = (mastered * 100)/exercise_count
+
+        context["students"].append({  # this could be DRYer
+            "first_name": user.first_name,
+            "last_name": user.last_name,
+            "username": user.username,
+            "name": user.get_name(),
+            "id": user.id,
+            "exercise_logs": log_table,
+            "progress": progress,
+            "mastered": mastered,
+            "struggling": struggling,
+            "mastery" : mastery,
+        })
+
+
+    progress = 0
+    mastered = 0
+    struggling = 0
+    for student in  context["students"]:
+        for ex in student["exercise_logs"]:
```

Comment:
Cleaner to loop over `student["exercise_logs"].values()` so you don't need to keep doing the lookup into `student["exercise_logs"]`